File: src/prod/src/Common/TimeoutHelper.cpp
// ------------------------------------------------------------
// Copyright (c) Microsoft Corporation.  All rights reserved.
// Licensed under the MIT License (MIT). See License.txt in the repo root for license information.
// ------------------------------------------------------------

#include "stdafx.h"

namespace Common
{

    TimeoutHelper::TimeoutHelper(TimeSpan timeout)
    {
        SetRemainingTime(timeout);
    }

    TimeSpan TimeoutHelper::GetRemainingTime() const
    {
        if (deadline_ == StopwatchTime::MaxValue)
        {
            return TimeSpan::MaxValue;
        }
        else
        {
            TimeSpan remaining = deadline_ - Stopwatch::Now();
            if (remaining <= TimeSpan::Zero)
            {
                return TimeSpan::Zero;
            }
            else
            {
                return remaining;
            }
        }
    }

    void TimeoutHelper::SetRemainingTime(TimeSpan timeout)
    {
        originalTimeout_ = timeout;
        deadline_ = Stopwatch::Now() + timeout;
    }
}

File: src/prod/src/managed/Api/src/System/Fabric/Management/common/TimeoutHelper.cs
// ------------------------------------------------------------
// Copyright (c) Microsoft Corporation.  All rights reserved.
// Licensed under the MIT License (MIT). See License.txt in the repo root for license information.
// ------------------------------------------------------------

namespace System.Fabric.Management.Common
{
    using System;
    using System.Fabric.Strings;

    internal class TimeoutHelper
    {        
        private DateTime deadline;

        public TimeoutHelper(TimeSpan timeout)
        {            
            deadline = (timeout == TimeSpan.MaxValue) ? DateTime.MaxValue : DateTime.Now.Add(timeout);
        }

        public TimeSpan GetRemainingTime()
        {
            if (deadline == DateTime.MaxValue)
            {
                return TimeSpan.MaxValue;
            }

            TimeSpan remaining = deadline - DateTime.Now;
            return remaining <= TimeSpan.Zero ? TimeSpan.Zero : remaining;
        }

        public void ThrowIfExpired()
        {
            if (this.GetRemainingTime() == TimeSpan.Zero)
            {
                throw new TimeoutException(StringResources.Error_OperationTimedOut);
            }
        }
    }
}